python/ray/train/tests/test_logging.py

```diff
@@ -0,0 +1,173 @@
+
+    assert "ham" not in capsys.readouterr().err
+
+
+def test_message_format(configure_logging, reset_logging, shutdown_only):
+    ray.init()
+    logger = logging.getLogger("ray.train.spam")
+
+    logger.info("ham")
+
+    log_path = os.path.join(get_log_directory(), "ray-train.log")
+    with open(log_path, "r") as f:
+        log_contents = f.read()
+    (
+        logged_ds,
+        logged_ts,
+        logged_level,
+        logged_filepath,
+        sep,
+        logged_msg,
+    ) = log_contents.split()
+
+    try:
+        datetime.strptime(f"{logged_ds} {logged_ts}", "%Y-%m-%d %H:%M:%S,%f")
+    except ValueError:
+        raise Exception(f"Invalid log timestamp: {logged_ds} {logged_ts}")
+
+    assert logged_level == logging.getLevelName(logging.INFO)
+    assert re.match(r"test_logging.py:\d+", logged_filepath)
+    assert logged_msg == "ham"
+
+
+def test_custom_config(reset_logging, monkeypatch, tmp_path):
+    config_path = tmp_path / "logging.yaml"
+    monkeypatch.setenv(LOG_CONFIG_PATH_ENV, config_path)
+
+    handlers = {
+        "console": {"class": "logging.StreamHandler", "stream": "ext://sys.stdout"}
+    }
+    loggers = {
+        "ray.train": {
+            "level": "CRITICAL",
+            "handlers": ["console"],
+        },
+    }
+    config = {
+        "version": 1,
+        "handlers": handlers,
+        "loggers": loggers,
+        "disable_existing_loggers": False,
+    }
+    with open(config_path, "w") as file:
+        yaml.dump(config, file)
+
+    configure_logging()
+
+    logger = logging.getLogger("ray.train")
+
+    assert logger.getEffectiveLevel() == logging.CRITICAL
+    assert len(logger.handlers) == 1
+    assert isinstance(logger.handlers[0], logging.StreamHandler)
+
+
+def test_json_logging_configuration(
+    capsys, reset_logging, monkeypatch, shutdown_only, propagate_logs
+):
+    import json
+
+    monkeypatch.setenv(LOG_ENCODING_ENV, "JSON")
+    ray.init()
+
+    configure_logging()
+
+    logger = logging.getLogger("ray.train")
+
+    # Ensure handlers correctly setup
+    handlers = logger.handlers
+    assert sum(handler.name == "file_json" for handler in handlers) == 1
+    assert sum(handler.name == "console" for handler in handlers) == 1
+
+    logger.info("ham")
+    logger.debug("turkey")
+
+    log_path = os.path.join(get_log_directory(), "ray-train.log")
+    with open(log_path) as file:
+        log_contents = file.read()
+
+    # Validate the log is in JSON format (a basic check for JSON)
+    messages = []
+    for log_line in log_contents.splitlines():
+        log_dict = json.loads(log_line)  # will error if not a json line
+        messages.append(log_dict["message"])
+
+    assert "ham" in messages
+    assert "turkey" in messages
+
+    # Validate console logs are in text mode
+    console_log_output = capsys.readouterr().err
+    for log_line in console_log_output.splitlines():
+        with pytest.raises(json.JSONDecodeError):
+            json.loads(log_line)
```

Comment:
So all `logger.warning(...)` gets logged in JSON format to the stdout as well? Does it make more sense for console outputs to be in the normal text format, and then only the `ray-train.log` contains the JSON format?

Also, is it printed to `stderr` or `stdout`?

Reply:
1. for `logger.warning(...)`, it is logged in JSON format to `ray-train.log` file, and logged to console as normal text format. The code here will raise `pytest.raises(json.JSONDecodeError)` because it is normal text, not json in the console output format.

2. By default, all the console output in python logger goes to `stderr`. Reference: https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial. and https://github.com/hongpeng-guo/ray/blob/6a5fc2d39b0265e7b578f069a84ae772c123801b/python/ray/_private/log.py#L30